_memory_viz.py: Visualize how blocks fit into segments.

zdevito opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #95753
* #95541
* #95357
* #95356
* __->__ #91336
* #96133

Add a segment_plot command that visualizes how blocks are allocated into segments.
This is similar to the 'stats' command but produces an interactive html viewer rather
than text dump, allowing exploration of stack traces.

It also adds the ability to see the layout at any point in the trace by starting from the
snapshot and then apply the events backwards to reconstruct what memory would have looked like.

Example:
![Screen Shot 2022-12-22 at 3 32 49 PM](https://user-images.githubusercontent.com/370202/209242650-b952372e-37ac-400a-a01c-13be2b5426fa.png)

zdevito committed:
_memory_viz.py: Visualize how blocks fit into segments.

Add a segment_plot command that visualizes how blocks are allocated into segments.
This is similar to the 'stats' command but produces an interactive html viewer rather
than text dump, allowing exploration of stack traces.

It also adds the ability to see the layout at any point in the trace by starting from the
snapshot and then apply the events backwards to reconstruct what memory would have looked like.

[ghstack-poisoned]

zdevito committed:
Update on "_memory_viz.py: Visualize how blocks fit into segments."


Add a segment_plot command that visualizes how blocks are allocated into segments.
This is similar to the 'stats' command but produces an interactive html viewer rather
than text dump, allowing exploration of stack traces.

It also adds the ability to see the layout at any point in the trace by starting from the
snapshot and then apply the events backwards to reconstruct what memory would have looked like.

Example:
![Screen Shot 2022-12-22 at 3 32 49 PM](https://user-images.githubusercontent.com/370202/209242650-b952372e-37ac-400a-a01c-13be2b5426fa.png)


[ghstack-poisoned]

zdevito committed:
Update on "_memory_viz.py: Visualize how blocks fit into segments."


Add a segment_plot command that visualizes how blocks are allocated into segments.
This is similar to the 'stats' command but produces an interactive html viewer rather
than text dump, allowing exploration of stack traces.

It also adds the ability to see the layout at any point in the trace by starting from the
snapshot and then apply the events backwards to reconstruct what memory would have looked like.

Example:
![Screen Shot 2022-12-22 at 3 32 49 PM](https://user-images.githubusercontent.com/370202/209242650-b952372e-37ac-400a-a01c-13be2b5426fa.png)


[ghstack-poisoned]

vadimkantorov commented:
It would be cool to also add a helper function that can output memory fragmentation and maybe some other statistics.

Then these metrics can be reported in the OOM exception message

zdevito committed:
Update on "_memory_viz.py: Visualize how blocks fit into segments."


Add a segment_plot command that visualizes how blocks are allocated into segments.
This is similar to the 'stats' command but produces an interactive html viewer rather
than text dump, allowing exploration of stack traces.

It also adds the ability to see the layout at any point in the trace by starting from the
snapshot and then apply the events backwards to reconstruct what memory would have looked like.

Example:
![Screen Shot 2022-12-22 at 3 32 49 PM](https://user-images.githubusercontent.com/370202/209242650-b952372e-37ac-400a-a01c-13be2b5426fa.png)


[ghstack-poisoned]

zdevito committed:
Update on "_memory_viz.py: Visualize how blocks fit into segments."


Add a segment_plot command that visualizes how blocks are allocated into segments.
This is similar to the 'stats' command but produces an interactive html viewer rather
than text dump, allowing exploration of stack traces.

It also adds the ability to see the layout at any point in the trace by starting from the
snapshot and then apply the events backwards to reconstruct what memory would have looked like.

Example:
![Screen Shot 2022-12-22 at 3 32 49 PM](https://user-images.githubusercontent.com/370202/209242650-b952372e-37ac-400a-a01c-13be2b5426fa.png)


[ghstack-poisoned]

zdevito committed:
Update on "_memory_viz.py: Visualize how blocks fit into segments."


Add a segment_plot command that visualizes how blocks are allocated into segments.
This is similar to the 'stats' command but produces an interactive html viewer rather
than text dump, allowing exploration of stack traces.

It also adds the ability to see the layout at any point in the trace by starting from the
snapshot and then apply the events backwards to reconstruct what memory would have looked like.

Example:
![Screen Shot 2022-12-22 at 3 32 49 PM](https://user-images.githubusercontent.com/370202/209242650-b952372e-37ac-400a-a01c-13be2b5426fa.png)


[ghstack-poisoned]

zdevito committed:
Update on "_memory_viz.py: Visualize how blocks fit into segments."


Add a segment_plot command that visualizes how blocks are allocated into segments.
This is similar to the 'stats' command but produces an interactive html viewer rather
than text dump, allowing exploration of stack traces.

It also adds the ability to see the layout at any point in the trace by starting from the
snapshot and then apply the events backwards to reconstruct what memory would have looked like.

Example:
![Screen Shot 2022-12-22 at 3 32 49 PM](https://user-images.githubusercontent.com/370202/209242650-b952372e-37ac-400a-a01c-13be2b5426fa.png)


[ghstack-poisoned]

zdevito committed:
Update on "_memory_viz.py: Visualize how blocks fit into segments."


Add a segment_plot command that visualizes how blocks are allocated into segments.
This is similar to the 'stats' command but produces an interactive html viewer rather
than text dump, allowing exploration of stack traces.

It also adds the ability to see the layout at any point in the trace by starting from the
snapshot and then apply the events backwards to reconstruct what memory would have looked like.

Example:
![Screen Shot 2022-12-22 at 3 32 49 PM](https://user-images.githubusercontent.com/370202/209242650-b952372e-37ac-400a-a01c-13be2b5426fa.png)


[ghstack-poisoned]

zdevito committed:
Update on "_memory_viz.py: Visualize how blocks fit into segments."


Add a segment_plot command that visualizes how blocks are allocated into segments.
This is similar to the 'stats' command but produces an interactive html viewer rather
than text dump, allowing exploration of stack traces.

It also adds the ability to see the layout at any point in the trace by starting from the
snapshot and then apply the events backwards to reconstruct what memory would have looked like.

Example:
![Screen Shot 2022-12-22 at 3 32 49 PM](https://user-images.githubusercontent.com/370202/209242650-b952372e-37ac-400a-a01c-13be2b5426fa.png)


[ghstack-poisoned]

zdevito committed:
Update on "_memory_viz.py: Visualize how blocks fit into segments."


Add a segment_plot command that visualizes how blocks are allocated into segments.
This is similar to the 'stats' command but produces an interactive html viewer rather
than text dump, allowing exploration of stack traces.

It also adds the ability to see the layout at any point in the trace by starting from the
snapshot and then apply the events backwards to reconstruct what memory would have looked like.

Example:
![Screen Shot 2022-12-22 at 3 32 49 PM](https://user-images.githubusercontent.com/370202/209242650-b952372e-37ac-400a-a01c-13be2b5426fa.png)


[ghstack-poisoned]

zdevito committed:
Update on "_memory_viz.py: Visualize how blocks fit into segments."


Add a segment_plot command that visualizes how blocks are allocated into segments.
This is similar to the 'stats' command but produces an interactive html viewer rather
than text dump, allowing exploration of stack traces.

It also adds the ability to see the layout at any point in the trace by starting from the
snapshot and then apply the events backwards to reconstruct what memory would have looked like.

Example:
![Screen Shot 2022-12-22 at 3 32 49 PM](https://user-images.githubusercontent.com/370202/209242650-b952372e-37ac-400a-a01c-13be2b5426fa.png)


[ghstack-poisoned]

zdevito committed:
Update on "_memory_viz.py: Visualize how blocks fit into segments."


Add a segment_plot command that visualizes how blocks are allocated into segments.
This is similar to the 'stats' command but produces an interactive html viewer rather
than text dump, allowing exploration of stack traces.

It also adds the ability to see the layout at any point in the trace by starting from the
snapshot and then apply the events backwards to reconstruct what memory would have looked like.

Example:
![Screen Shot 2022-12-22 at 3 32 49 PM](https://user-images.githubusercontent.com/370202/209242650-b952372e-37ac-400a-a01c-13be2b5426fa.png)


[ghstack-poisoned]

zdevito committed:
Update on "_memory_viz.py: Visualize how blocks fit into segments."


Add a segment_plot command that visualizes how blocks are allocated into segments.
This is similar to the 'stats' command but produces an interactive html viewer rather
than text dump, allowing exploration of stack traces.

It also adds the ability to see the layout at any point in the trace by starting from the
snapshot and then apply the events backwards to reconstruct what memory would have looked like.

Example:
![Screen Shot 2022-12-22 at 3 32 49 PM](https://user-images.githubusercontent.com/370202/209242650-b952372e-37ac-400a-a01c-13be2b5426fa.png)


[ghstack-poisoned]

zdevito committed:
Update on "_memory_viz.py: Visualize how blocks fit into segments."


Add a segment_plot command that visualizes how blocks are allocated into segments.
This is similar to the 'stats' command but produces an interactive html viewer rather
than text dump, allowing exploration of stack traces.

It also adds the ability to see the layout at any point in the trace by starting from the
snapshot and then apply the events backwards to reconstruct what memory would have looked like.

Example:
![Screen Shot 2022-12-22 at 3 32 49 PM](https://user-images.githubusercontent.com/370202/209242650-b952372e-37ac-400a-a01c-13be2b5426fa.png)


[ghstack-poisoned]

bhosmer commented on torch/cuda/_memory_viz.py:
```diff
@@ -824,6 +837,586 @@ def add_element(size, tensor_key, version):
+            t = ts[i]
+            if i + 1 < len(ts):
+                tnext = ts[i + 1]
+                if t['action'] == 'free_requested' and tnext['action'] == 'free_completed' and t['addr'] == tnext['addr']:
```
requested/completed pairs on the same address are guaranteed to be next to each other I guess?

zdevito replied:
They aren't always (sometimes the completed gets deferred), but that is the rare case, so it is nice to show most things as a single free event, and then highlight the rare cases by splitting the requested/completed events.

bhosmer commented on torch/cuda/_memory_viz.py:
```diff
@@ -824,6 +837,586 @@ def add_element(size, tensor_key, version):
+import {zoom, zoomIdentity} from "https://cdn.skypack.dev/d3-zoom@3";
+import {brushX} from "https://cdn.skypack.dev/d3-brush@3";
+
+let trace_data = $PLOT_DATA
```
I'm guessing this (and others below) could probably be `const` rather than `let` (anything that's never reassigned, even if you mutate object props, array items etc.) 

ADDING JAVASCRIPT VALUE

bhosmer reviewed (approved):
LGTM (caveat, I didn't actually load some generated html in a browser, but I didn't see anything on a read-through). Only real suggestion is a doc nit, totally ignorable unless you're in here again for some other reason

zdevito commented:
@pytorchbot merge

zdevito commented:
@pytorchbot merge
